### PR TITLE
improve error messages returned from dbus api

### LINKED ISSUE #567: Improve error messages returned from D-Bus API
Currently, some of the D-Bus error messages returned from BlueChi provide little to no information what went wrong so that the caller has a hard time figuring out what exactly went wrong, e.g. 

- https://github.com/containers/bluechi/blob/main/src/manager/manager.c#L662
Returns `Internal error` when the structure of the message is actually wrong. So returning `SD_BUS_ERROR_INVALID_ARGS` with a message like `Invalid arguments` would be easier to understand as caller. 
- https://github.com/containers/bluechi/blob/main/src/manager/monitor.c#L348
`Unsubscribe` returns only `Internal error` when it actually couldn't find the subscription to remove. 
- another example is the `Internal error` when listing units via `bluechictl` described in #568

*Acceptance criteria:*
Checking at least all public BlueChi API methods and changing the error type and message to provide a better understanding what went wrong.

### PR DESCRIPTION
**Pull request by Iiqbal2000**
Improving some error messages returned from the D-bus API

Fixes #567

### CODE REVIEW

**Comment by rhatdan** [comment]
Nice work @Iiqbal2000. thanks for your PR.
LGTM

**Comment by Iiqbal2000** [comment]
Let me know if there are any further changes needed 😉

**Comment by mwperina** [comment]
Hi @Iiqbal2000,
BlueChi has recently become part of Eclipse Foundation and as such each contributor to the project needs to have an account in Eclipse and sign Eclipse Contributor Agreement: https://www.eclipse.org/legal/ECA.php
Could you please take a achieve it? Without ECA confirmation we cannot accept your contribution.
Thanks a lot,
Martin

**Comment by Iiqbal2000** [comment]
Hi @mwperina, I have signed it

**Comment by mwperina** [comment]
Hi @Iiqbal2000,
ECA check is now green, thanks.
Now please take a look at the code formatting warnings, you check code formatting even on your local machine: https://github.com/containers/bluechi/blob/main/README.developer.md#code-style

File: src/agent/agent.c [hunk 1]
```diff
@@ -786,27 +786,27 @@ static int agent_method_get_unit_properties(sd_bus_message *m, void *userdata, U
...
         int r = sd_bus_message_read(m, "ss", &unit, &interface);
         if (r < 0) {
-                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_FAILED, "Internal Error");
+                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_FAILED, "Failed to read a message containing unit and interface: %s", strerror(-r));
         }
 
         r = assemble_object_path_string(SYSTEMD_OBJECT_PATH "/unit", unit, &unit_path);
         if (r < 0) {
-                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_FAILED, "Internal Error");
+                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_FAILED, "Out of memory");
```
Comment (engelmi) [inline src/agent/agent.c]: NIT: Could you change the return code to `SD_BUS_ERROR_NO_MEMORY` and state in the message when it went OOM, e.g. in this case `OOM when assembling unit path`.

**Comment by Iiqbal2000** [comment]
@mwperina could you recheck it?

File: src/libbluechi/socket.c [hunk 2]
```diff
@@ -5,13 +5,12 @@
...
 #include <unistd.h>
 
-#include "libbluechi/common/common.h"
 #include "socket.h"
 
 int create_tcp_socket(uint16_t port) {
         struct sockaddr_in6 servaddr;
 
-        _cleanup_fd_ int fd = socket(AF_INET6, SOCK_STREAM | SOCK_CLOEXEC | SOCK_NONBLOCK, 0);
+        int fd = socket(AF_INET6, SOCK_STREAM | SOCK_CLOEXEC | SOCK_NONBLOCK, 0);
```
Comment (engelmi) [inline src/libbluechi/socket.c]: Please revert this change and readd the `_cleanup_fd_` and the `seal_fd` below. It fixes a (possible) leaking fd (done in https://github.com/containers/bluechi/pull/602).

File: src/libbluechi/common/cfg.c [hunk 3]
```diff
@@ -326,7 +326,7 @@ int cfg_s_set_value(
                         return -ENOMEM;
                 }
         }
-        struct config_option *replaced = hashmap_set(
+        struct config_option *replaced = (struct config_option *) hashmap_set(
```
Comment (engelmi) [inline src/libbluechi/common/cfg.c]: What is the reason for adding these type casts? It isn't needed as far as I know, but I am wondering if this is related to https://github.com/containers/bluechi/issues/459 and removing the warnings we get?

File: src/libbluechi/socket.c (hunk 2 above)
Comment (mwperina) [inline src/libbluechi/socket.c]: Please revert this change, having `_cleanup_fd_` here is intentional: https://github.com/containers/bluechi/pull/602

File: src/libbluechi/socket.c [hunk 4]
```diff
@@ -39,7 +38,7 @@ int create_tcp_socket(uint16_t port) {
                 return -errsv;
         }
 
-        return steal_fd(&fd);
+        return fd;
```
Comment (mwperina) [inline src/libbluechi/socket.c]: Please revert this change, having `steal_fd` here is intentional: https://github.com/containers/bluechi/pull/602

**Review by mwperina — CHANGES_REQUESTED** [review]
In general looks good, just 2 small comments. 
Also  I'd probably squash both commits into a signle one. And please rebase your patch on top of latest main branch content.

**Review by engelmi — COMMENTED** [review]
I added a few comments. Could you also sign-off the second commit?
Overall really nice work!

File: src/libbluechi/common/cfg.c (hunk 3 above)
Comment (Iiqbal2000) [reply src/libbluechi/common/cfg.c]: Yep, it is related to this issue #459

Comment (Iiqbal2000) [reply src/libbluechi/common/cfg.c]: Do you think it should be reverted?

Comment (engelmi) [reply src/libbluechi/common/cfg.c]: I verified that your changes are indeed completely fixing #459 - nicely done! @Iiqbal2000 
Only the `hashmap.c` submodule would be needed to change to v0.7.0. 

> Do you think it should be reverted?

For this PR, yes please revert it. Since this is something unrelated to the dbus return message and a library version needs to change, it should go into a dedicated PR. 
So, I'd ask you to:
- move the hashmap.c changes to a new branch and create another PR
- revert the hashmap.c changes in this PR

Would you be up for that? If you need any support on the way, just reach out, of course :)

**Comment by Iiqbal2000** [comment]
@mwperina  @engelmi Could you check it again, I have fixed it

**Comment by Iiqbal2000** [comment]
The changes pertaining to hashmap.c requires me to update the version to v0.7.0 and proceed with this: "move the hashmap.c changes to a new branch and create another PR", correct? @engelmi

File: src/libbluechi/common/cfg.c (hunk 3 above)
Comment (Iiqbal2000) [reply src/libbluechi/common/cfg.c]: The changes pertaining to hashmap.c requires me to update the version to v0.7.0 and proceed with this: "move the hashmap.c changes to a new branch and create another PR", correct? @engelmi

Comment (engelmi) [reply src/libbluechi/common/cfg.c]: Yes, exactly (and update to version v0.7.0 in the new branch, of course). Thanks! @Iiqbal2000

**Review by mwperina — APPROVED** [review]
LGTM

File: src/agent/agent.c [hunk 5]
```diff
@@ -847,27 +862,36 @@ static int agent_method_get_unit_property(sd_bus_message *m, void *userdata, UNU
...
+                                m,
+                                SD_BUS_ERROR_FAILED,
+                                "Failed to read a message containing unit, interface, and property: %s",
+                                strerror(-r));
         }
 
         r = assemble_object_path_string(SYSTEMD_OBJECT_PATH "/unit", unit, &unit_path);
         if (r < 0) {
-                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_FAILED, "Internal Error");
+                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_FAILED, "Out of memory");
```
Comment (rhatdan) [inline src/agent/agent.c]: Should this be  SD_BUS_ERROR_NO_MEMORY?

File: src/agent/agent.c [hunk 6]
```diff
@@ -907,32 +931,44 @@ static int agent_method_set_unit_properties(sd_bus_message *m, void *userdata, U
...
+                                m,
+                                SD_BUS_ERROR_FAILED,
+                                "Failed to read a message containing unit and runtime: %s",
+                                strerror(-r));
         }
 
         r = assemble_object_path_string(SYSTEMD_OBJECT_PATH "/unit", unit, &unit_path);
         if (r < 0) {
-                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_FAILED, "Internal Error");
+                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_FAILED, "Out of memory");
```
Comment (rhatdan) [inline src/agent/agent.c]: SD_BUS_ERROR_NO_MEMORY?

File: src/agent/agent.c [hunk 7]
```diff
@@ -971,22 +1007,26 @@ static int agent_method_freeze_unit(sd_bus_message *m, void *userdata, UNUSED sd
...
         if (r < 0) {
-                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_INVALID_ARGS, "Invalid arguments");
+                return sd_bus_reply_method_errorf(
+                                m, SD_BUS_ERROR_INVALID_ARGS, "Invalid argument for the unit");
         }
 
         r = assemble_object_path_string(SYSTEMD_OBJECT_PATH "/unit", unit, &unit_path);
         if (r < 0) {
-                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_FAILED, "Internal Error");
+                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_FAILED, "Out of memory");
```
Comment (rhatdan) [inline src/agent/agent.c]: SD_BUS_ERROR_NO_MEMORY?

File: src/agent/agent.c [hunk 8]
```diff
@@ -1025,22 +1065,26 @@ static int agent_method_thaw_unit(sd_bus_message *m, void *userdata, UNUSED sd_b
...
```
Comment (rhatdan) [inline src/agent/agent.c]: SD_BUS_ERROR_NO_MEMORY?

File: src/manager/manager.c [hunk 9]
```diff
@@ -609,7 +613,7 @@ static int manager_method_list_units(sd_bus_message *m, void *userdata, UNUSED s
 
         req = malloc0_array(sizeof(*req), sizeof(req->sub_req[0]), manager->n_nodes);
         if (req == NULL) {
-                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_FAILED, "Internal error");
+                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_FAILED, "Out of memory");
```
Comment (rhatdan) [inline src/manager/manager.c]: SD_BUS_ERROR_NO_MEMORY?

**Comment by Iiqbal2000** [comment]
I have fixed it @rhatdan @mwperina @engelmi

**Review by engelmi — COMMENTED** [review]
LGTM
Before merging I think you have to sync the branch with main. Could you also merge the commits into one?

**Comment by Iiqbal2000** [comment]
> LGTM Before merging I think you have to sync the branch with main. Could you also merge the commits into one?

Done :)

**Comment by rhatdan** [comment]
I still see two commits?

**Comment by Iiqbal2000** [comment]
> I still see two commits?

Ohh sorry my bad

**Comment by Iiqbal2000** [comment]
It's done @rhatdan
